### PR TITLE
Ad9545 add model

### PR DESCRIPTION
**Pull request by tachicialex**
Add chip model for AD9545.

3f298a3a0f41 - pyadi-jiff: examples: Add AD9545
a0c9b3923e68 - pyadi-jiff: add model for AD9545
409488aa5ef8 - pyaadi-jiff: Add AD9545 support

### CODE REVIEW

File: adijif/clocks/ad9545.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        for i in range(0, 2):
+            for j in range(0, 4):
+                if self.input_refs[j] != 0 and self.PLL_used[i]:
+                    n_name = "n" + str(i) + "_profile_" + str(j)
+                    equations = equations + [self.config[n_name] * self.config["PLL_in_rate_" + str(j)] == self.config["PLL" + str(i) + "_rate"]]
+
+        self._add_equation(equations)
+
+        """ Instruct solver to try to avoid Minimum and Maximum PLL rates """
+        for i in range(0, 2):
```
Comment (tfcollins) [inline adijif/clocks/ad9545.py:245]: Please add a flag to disable optimizations/cost functions. For larger problems, these need to get converted to lexicographic versions

File: adijif/clocks/ad9545.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+    def _solve_gekko(self) -> bool:
+        """Local solve method for clock model.
+
+        Call model solver with correct arguments.
+
+        Returns:
+            bool: Always False
+        """
+        self.model.options.SOLVER = 1  # APOPT solver
```
Comment (tfcollins) [inline adijif/clocks/ad9545.py:321]: Do that default settings from clock.py not work for gekko?

File: examples/ad9545_example.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import adijif
+import pprint
+
+clk = adijif.ad9545(solver="gekko")
+
+input_refs = [1, 10e6, 0, 0]
+
+output_clocks = [30720000, 0, 0 , 0, 0, 0, 0 ,0 ,0, 0]
```
Comment (tfcollins) [inline examples/ad9545_example.py:8]: Do you have to put zeros? or can you just have a smaller list?

**Comment by tfcollins** [comment]
Please add a test for the AD9545. Something like this: https://github.com/analogdevicesinc/pyadi-jif/blob/main/tests/test_clocks.py#L10

File: examples/ad9545_example.py (hunk 3 above)
Comment (tachicialex) [reply examples/ad9545_example.py:8]: The zeros are necessary or i could change the method to accept an array of tuples (input_number, rate).

File: adijif/clocks/ad9545.py (hunk 2 above)
Comment (tachicialex) [reply adijif/clocks/ad9545.py:321]: With the default settings from clock.py , for some inputs i get solution not found.

**Comment by tachicialex** [comment]
Changelog V1 - > V2:
- added flags to disable cost functions
- removed clk_names arg.
- fixed multiple CPLEX objectives bug
- changed GEKKO solver_options after trying multiple usecases
- added tests for ad9545

File: examples/ad9545_example.py (hunk 3 above)
Comment (tfcollins) [reply examples/ad9545_example.py:10]: I would recommend accepting only clocks the user cares about aka not zero pad. This is how the other clock chips function. But also limit the user to 10 output clocks

File: tests/test_clocks.py [hunk 4]
```diff
@@ -6,6 +6,59 @@
...
+
+    input_refs = [1, 10e6, 0, 0]
+    output_clocks = [30720000, 0, 0 , 0, 0, 0, 0 ,0 ,0, 0]
+
+    input_refs = list(map(int, input_refs))  # force to be ints
+    output_clocks = list(map(int, output_clocks))  # force to be ints
+
+    clk.set_requested_clocks(input_refs, output_clocks)
+
+    clk.solve()
```
Comment (tfcollins) [inline tests/test_clocks.py:46]: Can you check the config against a reference output here

File: adijif/clocks/ad9545.py (hunk 2 above)
Comment (tfcollins) [reply adijif/clocks/ad9545.py:321]: If you update clocks.py to these settings do other tests fail?

Comment (tachicialex) [reply adijif/clocks/ad9545.py:321]: So i added this in clock.py: 
        self.model.options.SOLVER = 1  # APOPT solver
        self.model.solver_options = [
            "minlp_maximum_iterations 300000000",
            "minlp_max_iter_with_int_sol 6000",
            "minlp_as_nlp 0",
            "minlp_branch_method 1",
            "minlp_integer_tol 0.0001"
        ]
And when running 'make test' everything seems to be fine.

**Comment by tachicialex** [comment]
Changelog V2 -> V3:
- changed zero padded inputs to tuples: (ref_nr, rate_hz)
- when adding integer_var changed domain to normal min, max. For some reason
CPLEX would always return infeasible when using domains.
- added solution checks in test_ad9545_validate_pass

**Comment by tachicialex** [comment]
Changelog V3 -> V4:
- instead of raw solution comparison, check if rate formulas are ok
CPLEX will give different answers on different hardware

**Pull request merged**